2015/presentations/condemned_to_reinvent_sql/text.md

```diff
@@ -0,0 +1,342 @@
+    date = models.DateTimeField(null=False, blank=False)
+```
+
+The most obvious case where the ORM falls short, is any non-trivial
+aggregation. Let's say we would like to know which months are the most (or
+least) busy, per building, so we can plan construction work on the campus.
+
+With SQL, that's easy. First, fetch number of lectures happening in each
+building in each month:
+
+```sql
+select
+    "lectures_room"."building" as "building",
+    date_trunc('month', "lectures_lecture"."date") as "month",
+    count(*) as "count"
+from "lectures_lecture"
+left join "lectures_room" on "lectures_room"."id" = "lectures_lecture"."room_id"
+group by "building", "month"
+```
+
+result is going to be something like this:
+
+```
+ building |         month          | count
+----------+------------------------+-------
+ A2       | 2015-06-01 00:00:00+02 |   275
+ B3       | 2015-08-01 00:00:00+02 |   291
+ C3       | 2015-12-01 00:00:00+01 |   288
+ B4       | 2015-03-01 00:00:00+01 |   299
+ B2       | 2015-08-01 00:00:00+02 |   315
+ B1       | 2015-01-01 00:00:00+01 |   294
+ ...
+```
+
+Then, partition the result by building name, within each partition order rows
+by number of lectures, then fetch first and last value of "month" column:
+
+```sql
+with "busy_months" as (
+    ...
+)
+select distinct on ("building")
+    "building",
+    first_value("month") over "building_window" as "least_busy_month",
+    last_value("month")  over "building_window" as "most_busy_month"
+from
+    "busy_months"
+window "building_window" as (
+    partition by "building"
+    order by "count"
+    rows between unbounded preceding and unbounded following
+);
+```
+
+Which gives us the final report:
+
+```
+condemned=> select * from lectures_busy_months;
+ building |    least_busy_month    |    most_busy_month
+----------+------------------------+------------------------
+ A1       | 2015-02-01 00:00:00+01 | 2015-03-01 00:00:00+01
+ A2       | 2015-04-01 00:00:00+02 | 2015-01-01 00:00:00+01
+ A3       | 2015-02-01 00:00:00+01 | 2015-03-01 00:00:00+01
+ B1       | 2015-02-01 00:00:00+01 | 2015-03-01 00:00:00+01
+ B2       | 2015-02-01 00:00:00+01 | 2015-12-01 00:00:00+01
+ B3       | 2015-04-01 00:00:00+02 | 2015-07-01 00:00:00+02
+ B4       | 2015-02-01 00:00:00+01 | 2015-10-01 00:00:00+02
+ C1       | 2015-02-01 00:00:00+01 | 2015-12-01 00:00:00+01
+ C2       | 2015-01-01 00:00:00+01 | 2015-07-01 00:00:00+02
+ C3       | 2015-02-01 00:00:00+01 | 2015-07-01 00:00:00+02
+(10 rows)
+```
+
+Mapping this to the ORM is a bit of a problem, though:
+ * There is no `Building` model, so we can't say
+`Building.objects.annotate(count=Count(...))`
+ * There is no straightforward way to call `date_trunc` function and aggregate by that
+ * The ORM doesn't have any idea about `partition by` and `over` clauses.
+
+This doesn't mean that all hope is lost, though. Sane databases allow you to
+create a sort-of dynamic table out of that query. This is called a "view" -
```

Comment:
Personally I'd like to hear more about window function and common table expressions instead of views that seem to be well known to anyone that learned SQL a bit.